### ISSUE TITLE
[Bug][Compiler-v2] false positive error `mutable reference in local requires exclusive access but is borrowed`

### ISSUE THREAD

**Issue by rahxephon89** [aptos-labs/aptos-core#12849]
# 🐛 Bug

When compiling the following code using V2:

```
module 0x42::m {

    struct S has key, drop {
        g: G
    }
    struct G has copy, drop, store {
        id: u64,
    }
    fun foo(x: &S): &G {
        &x.g
    }

    fun foo_mut(_x: &mut S, _y:u64) {
    }

    fun id(g: &G):u64 {
        g.id
    }

    fun test2(
        s: S,
    ) {
        let x =   &mut s;
        let f = foo(x);
        foo_mut(x, id(f));
    }
}
```

The following error is generated
```
error: mutable reference in local `x` requires exclusive access but is borrowed
   │
37 │         let f = foo(x);
   │                 ------
   │                 │   │
   │                 │   previous freeze
   │                 used by call result
   ·
40 │         foo_mut(x, id(f));
   │                 ^  ----- conflicting reference `f` used here
   │                 │
   │                 requirement enforced here
```

No error will be generated when `test2` is defined as below:

```
    fun test2(
        s: S,
    ) {
        let x =   &mut s;
        let f = foo(x);
        let z = id(f);
        foo_mut(x, z);
    }
```

**Comment by rahxephon89** [comment]
The test `third_party/move/move-compiler/tests/move_check/translated_ir_tests/move/borrow_tests/eq_ok.move` has the same issue.

**Comment by rahxephon89** [comment]
fixed by #14276
